database/mysql_testdb.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2020 The Moov Authors
+// Use of this source code is governed by an Apache License
+// license that can be found in the LICENSE file.
+
+package database
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/moov-io/base/docker"
+	"github.com/moov-io/base/log"
+	"github.com/ory/dockertest"
+)
+
+var (
+	mysqlInstance *MySQLTestDB = &MySQLTestDB{}
+)
+
+func init() {
+	// after the above mysqlInstance is initialized with a zero-value
+	// write-lock our mysqlInstance so tests trying to read (RLock()
+	// are prevented until the underlying container is started.
+	mysqlInstance.mu.Lock()
+
+	go func() {
+		// ugly hack to stall .Wait() and allow tests to startup
+		time.Sleep(1 * time.Second)
+
+		// Block until all tests are finished
+		mysqlInstance.waiters.Wait()
+
+		// Shutdown our root connection and container
+		mysqlInstance.conn.Close()
+		mysqlInstance.container.Close()
+
+		// // Verify all connections are closed before closing DB
+		// if conns := r.DB.Stats().OpenConnections; conns != 0 {
+		// 	panic(fmt.Sprintf("found %d open MySQL connections", conns))
+		// }
+	}()
+}
+
+type MySQLTestDB struct {
```

Comment:
The idea of this thing is to act as a singleton for all `*testing.T` instances and create a new database whenever requested. We're blocking with an RWMutex that starts off in a write-locked state (so all reads are blocked) until the MySQL container is initialized. Then we're allowed to read-lock and initialize a database from `t.Name()`.

Reply:
As tests request a connection `TestMySQLConnection(t)` they're added into the `sync.WaitGroup` as the background goroutine (that does final cleanup) waits for them to finish. Once all registered tests are finished we close out the database.

This would be cleaner if Go had `BeforeAll` and `AfterAll` hooks we could register into, but without that we're doing a bit of racing to step in the middle of executing tests.

Reply:
A possible alternative: Use `TestMain(m *testing.M)` to run migrations then expose a method like `database.Cleanup(t *testing.T)` that will perform a teardown.

Having to call `database.Cleanup(t)` in the beginning of each test isn't too bad.